libdevice/include/asan_libdevice.hpp

```diff
@@ -0,0 +1,164 @@
+//==-- device-sanitizer-report.hpp - Structure and declaration for assert
+// support --==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#pragma once
+
+#include <cinttypes>
+
+#if OMP_LIBDEVICE && INTEL_COLLAB
+#pragma omp declare target
+#endif
+
+// NOTE This file should be sync with
+// unified-runtime/source/loader/layers/sanitizer/device_sanitizer_report.hpp
+
+enum class DeviceSanitizerErrorType : int32_t {
+  UNKNOWN,
+  OUT_OF_BOUNDS,
+  MISALIGNED,
+  USE_AFTER_FREE,
+  OUT_OF_SHADOW_BOUNDS,
+  UNKNOWN_DEVICE,
+  NULL_POINTER,
+};
+
+enum class DeviceSanitizerMemoryType : int32_t {
+  UNKNOWN,
+  USM_DEVICE,
+  USM_HOST,
+  USM_SHARED,
+  LOCAL,
+  PRIVATE,
+  MEM_BUFFER,
+  DEVICE_GLOBAL,
+};
+
+struct DeviceSanitizerReport {
+  int Flag = 0;
+
+  char File[256 + 1] = {};
+  char Func[256 + 1] = {};
+
+  int32_t Line = 0;
+
+  uint64_t GID0 = 0;
+  uint64_t GID1 = 0;
+  uint64_t GID2 = 0;
+
+  uint64_t LID0 = 0;
+  uint64_t LID1 = 0;
+  uint64_t LID2 = 0;
+
+  uintptr_t Address = 0;
+  bool IsWrite = false;
+  uint32_t AccessSize = 0;
+  DeviceSanitizerMemoryType MemoryType = DeviceSanitizerMemoryType::UNKNOWN;
+  DeviceSanitizerErrorType ErrorType = DeviceSanitizerErrorType::UNKNOWN;
+
+  bool IsRecover = false;
+};
+
+struct LocalArgsInfo {
+  uint32_t ArgIndex = 0;
+  uint64_t Size = 0;
+  uint64_t SizeWithRedZone = 0;
+};
+
+struct LaunchInfo {
+  uintptr_t PrivateShadowOffset =
+      0; // don't move this field, we use it in AddressSanitizerPass
+
+  uintptr_t LocalShadowOffset = 0;
+  uintptr_t LocalShadowOffsetEnd = 0;
+  DeviceSanitizerReport SanitizerReport;
+
+  uint32_t NumLocalArgs = 0;
+  LocalArgsInfo *LocalArgs = nullptr; // ordered by ArgIndex
+};
+
+constexpr unsigned ASAN_SHADOW_SCALE = 3;
+constexpr unsigned ASAN_SHADOW_GRANULARITY = 1ULL << ASAN_SHADOW_SCALE;
+
+// Based on the observation, only the last 24 bits of the address of the private
+// variable have changed, we use 31 bits(2G) to be safe.
+constexpr std::size_t ASAN_PRIVATE_SIZE = 0x7fffffffULL + 1;
+
+// These magic values are written to shadow for better error
+// reporting.
+constexpr int kUsmDeviceRedzoneMagic = (char)0x81;
```

Comment:
```suggestion
constexpr int kUsmDeviceRedzoneMagic = 0x81;
```
is the cast necessary?

Reply:
yes, it's necessary, or 0x81 will be treated as unsigned 0x81.